[RM Ch04] New Accelerator API Requirements chapter (#2593)
cntt-n/CNTT: doc/ref_model/chapters/chapter04.md

<p align="center"><b>Table 4-11:</b> Cloud Infrastructure management Resource Management Measurements</p>

<a name="4.1.7"></a>
### 4.1.7 Acceleration/Offload API Requirements 

> HW Accelerators and Offload functions with abstracted interfaces are preferred and can functionally be interchanged, but their characteristics might vary and it is likely that the CNFs/VNFs and Virtualization Infrastructure will have certification requirements for the different implementations. It is also often needed to have a SW implementation of these functions with the same abstracted interfaces for the deployment situations when there are no more HW Accelerator or Offload functions available.

Reviewer: ```suggestion
HW Accelerators and Offload functions with abstracted interfaces are preferred and can functionally be interchanged, but their characteristics might vary. It is also likely that the CNFs/VNFs and the Cloud Infrastructure will have certification requirements for the implementations.  An SW implementation of these functions is also often needed to have the same abstracted interfaces for the deployment situations when there are no more HW Accelerator or Offload functions available.
```

 
For Accelerators and Offload functions with externally exposed differences in their capabilities or management functionality these differences must be clear through the management API either explicit for the differing functions or implicit through the use of a unique APIs.
 

> Regardless of exposed or internal different capabilities and characteristics, the operators generally require a choice of multiple implementations also for Accelerators and Offload function realization, which drive the need for ease of portability in between implementations and vendors.

Reviewer 2: operators are telco operators here? Can you please clarify?

Author: Yes, it's the Telco Operators. I will improve the text to clarify that.

> The following table of requirement are derived from the need of the VNF/CNF applications, Cloud Infrastructure and Telco Operators to have multiple realizations of HW Acceleration and Offload functions that could also be implemented through a software implementation when no special hardware is available. These should be taken into any Reference Architecture to ensure that the different implementations on the market are as aligned as possible in their interfaces and that HW Acceleration and Offload gets an efficient eco system of accelerators that compete on their technical merits and not through obscured or proprietary interfaces.

Reviewer: ```suggestion
The following table of requirements are derived from the VNF/CNF applications, Cloud Infrastructure and Telco Operators needs to have multiple realizations of HW Acceleration and Offload functions that can also be implemented through SW when no special hardware is available. These requirements should be adopted in Reference Architectures to ensure that the different implementations on the market are as aligned as possible in their interfaces and that HW Acceleration and Offload functions get an efficient ecosystem of accelerators that compete on their technical merits and not through obscure or proprietary interfaces.
```

**Table 4-12** shows Acceleration/Offload API Capabilities.

<a name="Table4-12"></a>

| Ref       | Acceleration/Offload API Capability | Unit   | Definition/Notes                                                                            |
|-----------|--------------------------------------------|--------|---------------------------------------------------------------------------------------------|

> | e.api.001 | VNF/CNF usage of Accelerator standard i/f                           | Yes/No | VNF/CNF shall use abstracted standardized interfaces to the Acceleration/Offload function that enables multiple HW and SW implementation of the accelerated/offloaded function from multiple vendors in the Cloud Infrastructure                        |

Reviewer 2: Do you have any Accelerator standard i/f in mind in Lakelse timeframe?

Author: RM requirements intent and role is mainly to give the RA:s input of their more detailed requirements and specifications. We (RM) would not be in a position to drive the Lakelse standards adoption and realization roadmaps. Having said that, examples of existing protocols would be virtio, NVMe (emulating a physical drive in a SmartNIC/DPU/IPU), vSwitch offload (generally OVS) and some crypto both for in-line and look-aside encryption.
Then there are areas where a wanted standardization do not really have an existing standard on any level e.g. SR-IOV that neither have a higher layer accepted/adopted standard interface, e.g. how to support high performance secondary networking to the CNF applications. required by many Telco applications.

Reviewer 2: This would practically mean the introduction a set of gaps to the RA-s. If this is okay for RM I must accept it.

Author: Yes, that is correct.
This is inline with what we do in RM as well, where we try to follow the evolution (here of HW Accelerators) and apply the Anuket basic principles of decoupling of 1. Cloud Infrastructure implementations and VNF/CNF application implementations through rules, guidelines and interfaces as well as 2. HW Layer and Virtualization Layer inside the Cloud Infrastructure.
If these requirements on RM level are accepted for Lakelse, then they will be input for respective RA in next release to judge what are mature and implementable enough to be in that next release. 
If they can be described as Gaps already in the Lakelse RA-1 and RA-2 documents that would naturally be good, but probably not necessary.

> | e.api.001 | VNF/CNF usage of Accelerator standard i/f                           | Yes/No | VNF/CNF shall use abstracted standardized interfaces to the Acceleration/Offload function that enables multiple HW and SW implementation of the accelerated/offloaded function from multiple vendors in the Cloud Infrastructure                        |

Reviewer: ```suggestion
| e.api.001 | VNF/CNF usage of Accelerator standard i/f                           | Yes/No | VNF/CNF shall use abstracted standardized interfaces to the Acceleration/Offload function that enables HW and SW implementations of the accelerated/offloaded functions from multiple vendors in the Cloud Infrastructure.                        |
```

> | e.api.002 | Virtualisation Infrastructure SW usage of Accelerator standard i/f  | Yes/No | Virtualisation Infrastructure SW shall use abstracted standardised interfaces to the HW-Acceleration/Offload  function that enables multiple HW and SW implementation in the HW Infrastructure Layer of the accelerated function from multiple vendors  |

Reviewer: ```suggestion
| e.api.002 | Virtualisation Infrastructure SW usage of Accelerator standard i/f  | Yes/No | Virtualisation Infrastructure SW shall use abstracted standardised interfaces to the HW-Acceleration/Offload function. This would enable multiple HW and SW implementations in the HW Infrastructure Layer of the accelerated functions from multiple vendors.  |
```

Author: @pgoyal01 Text is corrected in a slightly different way, but hopefully to your approval.

> | e.api.003 | Accelerators offering standard i/f to HW Infra Layer                | Yes/No | Acceleration/Offload functions shall offer abstracted standardised interfaces for the Virtualisation Infrastructure and VNF/CNF applications                                                                                                             |

Reviewer: ```suggestion
| e.api.003 | Accelerators offering standard i/f to HW Infra Layer                | Yes/No | Acceleration/Offload functions shall offer abstracted standardised interfaces for the Virtualisation Infrastructure and VNF/CNF applications.                                                                                                             |
```

> | e.api.003 | Accelerators offering standard i/f to HW Infra Layer                | Yes/No | Acceleration/Offload functions shall offer abstracted standardized interfaces for the Virtualization Infrastructure and VNF/CNF applications                                                                                                             |

Reviewer: Can we combine the first 3 into a single requirement?

| e.api.001 | Accelerators offer standard i/f                 | Yes/No |  The Platform shall offer abstracted standardized interfaces for LCM of the Acceleration/Offload functions for use by the Cloud Infrastructure, VNF/CNF applications, and other users. |   

or something similar

Author: I have a hard time combining all of them, since they are different requirements on different entities.  Possibly 001 and 002 could be combined, but they will then need to be acted upon by different Anuket Groups (RI-x/RC-x and RA-x) and implementations (VNFs/CNFs as being the applications and Virtualizaton Infrastructure as being the Infrastructure internal layering requirements).
001: Mandate VNF/CNF to use abstracted API where they exist
002 Mandate SW Virtualization layer implementations (i.e. OpenStack and K8s) to use abstracted API where they exist
To your suggested text, it is far more than LCM interfaces we talk about here.
Why is it important to bundle or abstract the requirements? The risk is just that it become obfuscated.

Reviewer: @TFredberg Can we simplify? A suggestion made in reqt. 001. BTW we use the word "must".

Author: You are suggesting other requirements through the change of text. So, yes, that points to that we must change the text, since it is not properly understood. 
Usage of the word Platform neither helps me, since that is an undefined entity here.

Can we first agree on what the requirement is, and then we can possibly word-smith it into a simple enough form that still is correct.
001: Intend to say that VNFs and CNFs shall be written and tested to support abstracted Acceleration/Offload "standard" interface when such once are available, even if there exist vendor or device-specific interfaces that might be more comprehensive or efficient. It shall be up to an application-external orchestrator which of the application supported accelerators or offload functions to actually deploy at any point of deployment.
002: Intend to say the same as 001 but shall be valid for the Virtualization SW inside the Cloud Infrastructure, which is just one of the internal layers inside the Cloud Infrastructure.

> | e.api.004 | Accelerators offering virtualised functions                         | Yes/No | Acceleration/Offload functions for VNFs/CNFs should preferably be virtualised to allow multiple VNFs/CNFs to use the same Acceleration/Offload instance                                                                                                 |

Reviewer: ```suggestion
| e.api.004 | Accelerators offering virtualised functions                         | Yes/No | Acceleration/Offload functions for VNFs/CNFs should be virtualised to allow multiple VNFs/CNFs to use the same Acceleration/Offload instance.                                                                                                 |
```

> | e.api.005 | VNF/CNF Accelerator management functions access rights              | Yes/No | VNF/CNF management functions shall be capable to request Acceleration/Offload invocation without having elevated access rights                                                                                                                       |

Reviewer: I think "capable" implies the responsibility is on the "management functions"
```suggestion
| e.api.005 | VNF/CNF Accelerator management functions access rights              | Yes/No | VNF/CNF management functions shall be able to request Acceleration/Offload invocation without requiring elevated access rights.                                                                                                                       |
```

> | e.api.006 | Accelerators offering standard i/f to VNF/CNF management            | Yes/No | VNF/CNF management functions should be capable to request Acceleration/Offload invocation through abstracted standardised Management interfaces                                                                                                       |

Reviewer: ```suggestion
| e.api.006 | Accelerators offering standard i/f to VNF/CNF management            | Yes/No | VNF/CNF management functions should be able to request Acceleration/Offload invocation through abstracted standardised Management interfaces.                                                                                                       |
```

> | e.api.006 | Accelerators offering standard i/f to VNF/CNF management            | Yes/No | VNF/CNF management functions should be capable to request Acceleration/Offload invocation through abstracted standardized Management interfaces                                                                                                       |

Reviewer 2: Would this mean Cyborg in OpenStack and standardised API schemes for Device Plugins in Kubernetes?

Author: To my knowledge (but I am not an expert here) Cyborg is most a discovery and mapping interface for some resource orchestrator, and that is a possible API that could make sense in the management API area of the technology dependent RA-1.
The Device Plug-In I would think (again, I am no expert in this area) can abstract underlying implementation differences, so a specification of what a Device Plug-In can do and how to interface/manage Device Plug-Ins would be an example of an architecture dependent API that should be standardized in RA-2.

Reviewer: @TFredberg @CsatariGergely Again not an expert but Cyborg supports LCM for accelerators. It supports a subset of compute (Nova) operations for servers (in OSTK -- physical, VM or container) with accelerators.  So as to avoid flavor explosion, Cyborg uses the concept of device profiles -- a sort of "flavor" for devices -- that can be included in regular OSTK flavor definitions. Another distinction is an [abstraction of drivers](https://docs.openstack.org/cyborg/latest/user/architecture.html) which then interacts with specific vendor drivers -- so an abstract driver for FPGA, another for GPU, etc.

Reviewer 2: In my reading e.api.006 is about the management of accelerators and not the consumption of them. I also agree with @pgoyal01 in that Cyborg is more than a simple discovery and mapping interface. I can reach out for the Cyborg community in OpenStack to have a more experienced opinion. 
Device plugins are already [required in RA2](https://cntt.readthedocs.io/en/latest/ref_arch/kubernetes/chapters/chapter03.html?highlight=device%20plugin#device-plugin-framework). The issue with device plugins is that every device plugin offers a different API, so "device plugin" does not mean a standard interface to manage accelerators, it means a standard approach. If we would like to have a standard api we would need to specify a scheme for it and influence device plugins to use the scheme.

Author: The intent of e.api.006 is to have a decoupling in between the HW realization of the HW-accelerator (inside the Cloud Infrastructure) and the VNF/CNF application or its management function (like a resource orchestrator). To achieve that, there is a need to have a requirement on the Cloud Infrastructure exposed control/management interfaces. If that is fulfilled by the HW accelerator or the Virtualization Layer (OpenStack or K8s) is up to the Cloud Infrastructure realization. It is likely that the vendors of HW accelerators do not have vendor agnostic portable interfaces at least not in their early releases, and then it will be up to the Virtualization Layer to try their best to abstract the differences of the HW accelerators.

Author: Would agree on the discussion, but I think it will be a subject for respective RA rather than in the RM.

> | e.api.007 | VNFs/CNFs and Virtualisation Infrastructure Accelerator portability | Yes/No | VNFs/CNFs and Virtualisation Infrastructure SW should be designed to handle multiple types of Accelerator or Offload Function realisations even when their differences are exposed to the infrastructure or applications layers                    |

Reviewer: ```suggestion
| e.api.007 | VNFs/CNFs and Virtualisation Infrastructure Accelerator portability | Yes/No | VNFs/CNFs and Virtualisation Infrastructure SW should be designed to handle multiple types of Accelerator or Offload Function realisations even when their differences are exposed to the infrastructure or applications layers.                    |
```

> | e.api.007 | VNFs/CNFs and Virtualization Infrastructure Accelerator portability | Yes/No | VNFs/CNFs and Virtualization Infrastructure SW should be designed to handle multiple types of Accelerator or Offload Function realizations even when their differences are exposed to the infrastructure or applications layers                    |

Reviewer 2: Can you please clarify this requirement? Does this mean that each VNF/CNF should support a multiple types of accelerators, which are supported by the infrastructures?

Author: I can not know exactly which requirement text you are asking/commenting on, but I guess it is 007 and will answer as if that is the case:
Yes, an application and a Virtualization Infrastructure should be designed to make use of multiple different interfaces towards accelerators (and at least one SW implementation) when there are more common accelerators on the market. This is normally done through application certification and enable the applications to run on Cloud Infrastructure platforms that only have one or a selection of accelerators  available.
This is the norm in the IT and PC acceleration industry with GPU support being a good example.
The requirement do not say that  the situation of having vendor or chip specific interfaces is good, but recognize that this is often the case.

Reviewer 2: The comment is related to e.api.007 (line 288).
Thanks for the answer, but does this leave the discovery of available accelerators and self-configuration to use one (or more) particular accelerator(s) to the application? What happens if the application gets migrated to a different host with different set of accelerators?

Author: This requirement do not define who would make a selection of what of the available accelerators that will actually be used, when there are more than one. In my view that should be an orchestrator (or even more specified a resource orchestrator), but it technically could also be the VNF/CNF. In my view the selection of where to use the available HW accelerators should be in the hands of an orchestrator that can optimize the actual deployment in the actual situation, honoring the needed characteristics.

> | e.api.008 | VNFs/CNFs and Virtualisation Infrastructure Accelerator flexibility | Yes/No | VNFs/CNFs and Virtualisation Infrastructure SW shall be able to use any assigned instance and type of Accelerator or Offload Function that it is certified for                                                                                          |

Reviewer: ```suggestion
| e.api.008 | VNFs/CNFs and Virtualisation Infrastructure Accelerator flexibility | Yes/No | VNFs/CNFs and Virtualisation Infrastructure SW shall be able to use any assigned instance and type of Accelerator or Offload Function that they are certified for.                                                                                          |
```

 
The following table of requirements are derived from the VNF/CNF applications, Cloud Infrastructure and Telco Operators needs to have multiple realisations of HW Acceleration and Offload functions that can also be implemented through SW when no special hardware is available. These requirements should be adopted in Reference Architectures to ensure that the different implementations on the market are as aligned as possible in their interfaces and that HW Acceleration and Offload functions get an efficient ecosystem of accelerators that compete on their technical merits and not through obscure or proprietary interfaces.

**Table 4-12** shows Acceleration/Offload API Capabilities.

<a name="Table4-12"></a>

| Ref       | Acceleration/Offload API Capability | Unit   | Definition/Notes                                                                            |
|-----------|--------------------------------------------|--------|---------------------------------------------------------------------------------------------|

> | e.api.001 | VNF/CNF usage of Accelerator standard i/f                           | Yes/No | VNF/CNF shall use abstracted standardised interfaces to the Acceleration/Offload function that enables HW and SW implementations of the accelerated/offloaded functions from multiple vendors in the Cloud Infrastructure.                        |

Reviewer: ```suggestion
| e.api.001 | VNF/CNF usage of Accelerator standard i/f                           | Yes/No | VNF/CNF shall use abstracted standardised interfaces to the Acceleration/Offload functions. This would enable use of HW and SW implementations of the accelerated/offloaded functions from multiple vendors in the Cloud Infrastructure.                        |
```
